### PR TITLE
feat(tests): add JSON Schema test coverage

### PR DESCRIPTION
**Pull request by jeremyfiel**
This PR adds schema test coverage for the JSON Schema, 

this is based on the work in the OAS and Overlay projects by @MikeRalphson and @ralfhandl 

The schema tests do not cover JSON Schema 2020-12 keywords as defined by the `inputs` keyword. All Arazzo specific keywords are covered by the tests.

I've also included markdownlint-cli to clean up the spec file.

### CODE REVIEW

**Comment by jeremyfiel** [comment]
rebased on v1.0.0-dev to resolve conflicts.

**Review by ralfhandl — COMMENTED** [review]
@jeremyfiel Why is the specification markdown source changed by this PR?

**Comment by jeremyfiel** [comment]
Markdown lint was added and the rebase onto v1.0.0-dev included some recent updates by Frank   

I didn't change any wording on it

**Comment by jeremyfiel** [comment]
* stripped out the markdown lint updates
* removed the json version of the schema

**Review by jeremyfiel — COMMENTED** [review]
mistakenly removed the original `validate-markdown.yaml` with mdv checks.

File: scripts/format-markdown.sh
Comment (ralfhandl) [inline scripts/format-markdown.sh:1]: This file should be removed from this PR and only part of #291

Comment (jeremyfiel) [reply scripts/format-markdown.sh:1]: done

**Comment by jeremyfiel** [comment]
opps. somehow the validate-markdown.yaml was deleted again. added it back.

File: schemas/v1.0/readme.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# OpenAPI Arazzo 1.0.x JSON Schema
+
+Here you can find the JSON Schema for validating Arazzos of versions 1.0.x.
```
Comment (frankkilcommins) [inline schemas/v1.0/readme.md]:
```suggestion
Here you can find the JSON Schema for validating Arazzo Documents conforming to versions 1.0.x.
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+          "products": "{$inputs.products}"
+        }
+    successCriteria:
+      - condition: $statusCode == 200
+    onSuccess:
+      - name: existingCustomerNotEligible
+        type: end
+        criteria:
+          - condition: $statusCode == 200
+          - condition: $response.body.existingCustomerNotEligible == false
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml]:
```suggestion
          - condition: $response.body#/existingCustomerNotEligible == false
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        type: end
+        criteria:
+          - condition: $statusCode == 200
+          - context: $response.body
+            type: jsonpath
+            condition: $[?count(@.products) == 0]
+    outputs:
+      eligibilityCheckRequired: $response.body.eligibilityCheckRequired
+      eligibleProducts: $response.body.products
+      totalLoanAmount: $response.body.totalAmount
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml]:
```suggestion
      eligibilityCheckRequired: $response.body#/eligibilityCheckRequired
      eligibleProducts: $response.body#/products
      totalLoanAmount: $response.body#/totalAmount
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        type: goto
+        stepId: initiateBnplTransaction
+        criteria:
+          - condition: $statusCode == 201
+      - name: customerIsNotEligible
+        type: end
+        criteria:
+          - condition: $statusCode == 200
+    outputs:
+      customer: $response.body.links.self
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml]:
```suggestion
      customer: $response.body#/links/self
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+          "products": "{$steps.checkLoanCanBeProvided.outputs.eligibleProducts}"
+        }
+    successCriteria:
+      - condition: $statusCode == 202
+    onSuccess:
+      - name: CustomerAuthorizationRequired
+        type: goto
+        stepId: authenticateCustomerAndAuthorizeLoan
+        criteria:
+          - condition: $response.body.redirectAuthToken != null
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml]:
```suggestion
          - condition: $response.body#/redirectAuthToken != null
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+      - name: CustomerAuthorizationNotRequired
+        type: goto
+        stepId: retrieveFinalizedPaymentPlan
+        criteria:
+          - condition: $response.body.redirectAuthToken == null
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml]:
```suggestion
          - condition: $response.body#/redirectAuthToken == null
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    outputs:
+      redirectAuthToken: $response.body.redirectAuthToken
+      loanTransactionResourceUrl: $response.body.links.self
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml:215]:
```suggestion
      redirectAuthToken: $response.body#/redirectAuthToken
      loanTransactionResourceUrl: $response.body#/links/self
```

File: tests/v1.0/pass/bnpl-example.yaml [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+      - name: loanTransactionId
+        in: path
+        value: $steps.initiateBnplTransaction.outputs.loanTransactionId
+    requestBody:
+      contentType: application/json
+      payload: { "loanStatus": "Completed" }
+    successCriteria:
+      - condition: $statusCode == 204
+  outputs:
+    finalizedPaymentPlan: $steps.retrieveFinalizedPaymentPlan.finalizedPaymentPlan
```
Comment (frankkilcommins) [inline tests/v1.0/pass/bnpl-example.yaml]:
```suggestion
    finalizedPaymentPlan: $steps.retrieveFinalizedPaymentPlan.outputs.finalizedPaymentPlan
```

File: tests/v1.0/pass/oauth-example.yaml [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            refresh_token: $steps.do-the-auth-flow.outputs.my_refresh_token
+        successCriteria:
+          - condition: $statusCode == 200
+          - context: $response.body
+            condition: $.access_token != null
+            type: jsonpath
+        outputs:
+          access_token: $response.body.access_token
+          refresh_token: $response.body.refresh_token
+          expires_in: $response.body.expires_in
```
Comment (frankkilcommins) [inline tests/v1.0/pass/oauth-example.yaml]:
```suggestion
          access_token: $response.body#/access_token
          refresh_token: $response.body#/refresh_token
          expires_in: $response.body#/expires_in
```

File: tests/v1.0/pass/oauth-example.yaml [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            client_id: $inputs.client_id
+            client_secret: $inputs.client_secret
+            grant_type: client_credentials
+        successCriteria:
+          - condition: $statusCode == 200
+          - context: $response.body
+            condition: $.access_token != null
+            type: jsonpath
+        outputs:  
+          access_token: $response.body.access_token
```
Comment (frankkilcommins) [inline tests/v1.0/pass/oauth-example.yaml]:
```suggestion
          access_token: $response.body#/access_token
```

File: tests/v1.0/pass/oauth-example.yaml [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        - name: state
+          in: query
+          value: '12345'
+        successCriteria:
+          - condition: $statusCode == 200
+          - context: $response.body
+            condition: $.access_token != null
+            type: jsonpath
+        outputs:  
+          code: $response.body.code  # Not really, this is a query parameter
```
Comment (frankkilcommins) [inline tests/v1.0/pass/oauth-example.yaml]:
```suggestion
          code: $response.body#/code  # Not really, this is a query parameter
```

File: tests/v1.0/pass/oauth-example.yaml [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+            client_secret: $inputs.client_secret
+        successCriteria:
+          - condition: $statusCode == 200
+          - context: $response.body
+            condition: $.access_token != null
+            type: jsonpath
+        outputs:  
+          access_token: $response.body.access_token
+          refresh_token: $response.body.refresh_token
+          expires_in: $response.body.expires_in
```
Comment (frankkilcommins) [inline tests/v1.0/pass/oauth-example.yaml]:
```suggestion
          access_token: $response.body#/access_token
          refresh_token: $response.body#/refresh_token
          expires_in: $response.body#/expires_in
```

File: tests/v1.0/pass/pet-coupons-example.yaml [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+      - stepId: find-pet
+        operationId: findPetsByTags
+        parameters:
+          - name: pet_tags
+            in: query
+            value: $inputs.my_pet_tags
+        successCriteria:
+          - condition: $statusCode == 200
+        outputs:
+          my_pet_id: $outputs[0].id
```
Comment (frankkilcommins) [inline tests/v1.0/pass/pet-coupons-example.yaml]:
```suggestion
          my_pet_id: $response.body#/0/id
```

File: tests/v1.0/pass/pet-coupons-example.yaml [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        description: Find a coupon available for the selected pet.
+        operationId: getPetCoupons
+        parameters:
+          - name: pet_id
+            in: path
+            value: $steps.find-pet.outputs.my_pet_id
+        successCriteria:
+          - condition: $statusCode == 200
+        outputs:
+          my_coupon_code: $response.body.couponCode
```
Comment (frankkilcommins) [inline tests/v1.0/pass/pet-coupons-example.yaml]:
```suggestion
          my_coupon_code: $response.body#/couponCode
```

File: tests/v1.0/pass/pet-coupons-example.yaml [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        workflowId: place-order
+        parameters:
+          - name: pet_id
+            value: $steps.find-pet.outputs.my_pet_id
+          - name: coupon_code
+            value: $steps.find-coupons.outputs.my_coupon_code
+        successCriteria:
+          - condition: $statusCode == 200
+        outputs:
+          my_order_id: $workflow_order_id
```
Comment (frankkilcommins) [inline tests/v1.0/pass/pet-coupons-example.yaml]:
```suggestion
          my_order_id: $outputs.workflow_order_id
```

File: tests/v1.0/pass/pet-coupons-example.yaml [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            in: query
+            value: "available"
+          - reference: $components.parameters.page
+            value: 1
+          - reference: $components.parameters.pageSize
+            value: 10
+        successCriteria:
+          - condition: $statusCode == 200
+        outputs:
+          my_pet_id: $outputs[0].id
```
Comment (frankkilcommins) [inline tests/v1.0/pass/pet-coupons-example.yaml]:
```suggestion
          my_pet_id: $response.body#/0/id
```

File: tests/v1.0/pass/pet-coupons-example.yaml [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+      - stepId: place-order
+        workflowId: place-order
+        parameters:
+          - name: pet_id
+            value: $steps.find-pet.outputs.my_pet_id
+        successCriteria:
+          - condition: $statusCode == 200
+        outputs:
+          my_order_id: $workflow_order_id
```
Comment (frankkilcommins) [inline tests/v1.0/pass/pet-coupons-example.yaml]:
```suggestion
          my_order_id: $outputs.workflow_order_id
```

File: tests/v1.0/pass/pet-coupons-example.yaml [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+          payload:
+            petId: $inputs.pet_id
+            quantity: $inputs.quantity
+            couponCode: $inputs.coupon_code
+            status: placed
+            complete: false
+        successCriteria:
+          - condition: $statusCode == 200
+        outputs:
+          step_order_id: $response.body.id
```
Comment (frankkilcommins) [inline tests/v1.0/pass/pet-coupons-example.yaml]:
```suggestion
          step_order_id: $response.body#/id
```

**Review by frankkilcommins — CHANGES_REQUESTED** [review]
Thanks @jeremyfiel - i'm requested changed on the examples to ensure they are valid (which is being fixed for the examples directly in #260 )

Can you also take a look at the merge conflicts?

**Comment by jeremyfiel** [comment]
tbh, i'm not sure why this is a conflict. the file currently exists on `v1.0.0-dev` and is requested to be deleted in this PR. so it shouldn't conflict with anything.

**Comment by jeremyfiel** [comment]
interesting. adding the file back removes the conflict. I can create a separate pr to remove the file after this is merged.. lmk if that works

**Comment by jeremyfiel** [comment]
taking inspiration from @ralfhandl  to update the schema test coverage threshold to 100% like [this](https://github.com/OAI/Overlay-Specification/pull/109/files) will not work for this repo because we are not testing the JSON Schema 2020-12 keywords used in `inputs`.  I'm not sure if we want to peg it to ~63% coverage to indicate the threshold is met or not.  To be discussed...

**Review by frankkilcommins — APPROVED** [review]

**Pull request merged**